Services/JWSProvider.php

```diff
@@ -0,0 +1,192 @@
+<?php
+
+namespace Lexik\Bundle\JWTAuthenticationBundle\Services;
+
+use Lexik\Bundle\JWTAuthenticationBundle\Services\KeyLoader\KeyLoaderInterface;
+use Namshi\JOSE\JWS;
+
+/**
+ * JWS Provider, Namshi\JOSE library integration.
+ * Supports OpenSSL and phpseclib encryption engines.
+ *
+ * @internal
+ *
+ * @since 2.0
+ *
+ * @author Robin Chalas <robin.chalas@gmail.com>
+ */
+class JWSProvider implements JWSProviderInterface
+{
+    /**
+     * @var KeyLoaderInterface
+     */
+    protected $keyLoader;
+
+    /**
+     * @var string
+     */
+    protected $encryptionEngine;
+
+    /**
+     * @var string
+     */
+    protected $encryptionAlgorithm;
+
+    /**
+     * @var JWS
+     */
+    protected $jws;
+
+    /**
+     * @var string
+     */
+    protected $status;
+
+    /**
+     * @param KeyLoaderInterface $keyLoader
+     * @param string             $encryptionEngine
+     * @param string             $encryptionAlgorithm
+     *
+     * @throws \InvalidArgumentException If the given algorithm is not supported.
+     */
+    public function __construct(KeyLoaderInterface $keyLoader, $encryptionEngine, $encryptionAlgorithm)
+    {
+        $encryptionEngine = $encryptionEngine == 'openssl' ? 'OpenSSL' : 'SecLib';
+
+        if (!$this->isAlgorithmSupportedForEngine($encryptionEngine, $encryptionAlgorithm)) {
+            throw new \InvalidArgumentException(
+                sprintf('The algorithm "%s" is not supported for %s', $encryptionAlgorithm, $encryptionEngine)
+            );
+        }
+
+        $this->keyLoader           = $keyLoader;
+        $this->encryptionEngine    = $encryptionEngine;
+        $this->encryptionAlgorithm = $encryptionAlgorithm;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function createSignedToken(array $payload)
+    {
+        $this->jws = new JWS(['alg' => $this->encryptionAlgorithm], $this->encryptionEngine);
+        $this->jws->setPayload($payload);
+        $this->sign();
+
+        return $this;
```

Comment:
I'll start from this, maybe you'll be able to help me more by seeing it in action.
Thank you for the try :)